helix-core/src/main/java/org/apache/helix/HelixPropertyFactory.java

```diff
@@ -80,20 +80,27 @@ public HelixManagerProperty getHelixManagerProperty(String zkAddress, String clu
    * @param clusterName
    * @return
    */
+
   public static CloudConfig getCloudConfig(String zkAddress, String clusterName) {
+    return getCloudConfig(zkAddress, clusterName, null);
+  }
+  public static CloudConfig getCloudConfig(String zkAddress, String clusterName,
+      RealmAwareZkClient.RealmAwareZkConnectionConfig realmAwareZkConnectionConfig) {
     CloudConfig cloudConfig;
     RealmAwareZkClient dedicatedZkClient = null;
     try {
       if (Boolean.getBoolean(SystemPropertyKeys.MULTI_ZK_ENABLED) || zkAddress == null) {
         // If the multi ZK config is enabled or zkAddress is null, use realm-aware mode with
         // DedicatedZkClient
         try {
-          RealmAwareZkClient.RealmAwareZkConnectionConfig connectionConfig =
-              new RealmAwareZkClient.RealmAwareZkConnectionConfig.Builder()
-                  .setRealmMode(RealmAwareZkClient.RealmMode.SINGLE_REALM)
-                  .setZkRealmShardingKey("/" + clusterName).build();
+          if (realmAwareZkConnectionConfig == null) {
```

Comment:
I'm not sure if this is the right pattern. I would do
```
public static CloudConfig getCloudConfig(String zkAddress, String clusterName) {
    realmAwareZkConnectionConfig =
                new RealmAwareZkClient.RealmAwareZkConnectionConfig.Builder()
                    .setRealmMode(RealmAwareZkClient.RealmMode.SINGLE_REALM)
                    .setZkRealmShardingKey("/" + clusterName).build();
    return getCloudConfig(zkAddress, clusterName, realmAwareZkConnectionConfig);
  }
```
Instead of a null check here.

Reply:
TFTR. When we add an API to allow user to pass in `RealmAwareZkConnectionConfig` and when user pass in a null obj, shouldn't we create a default one instead of using the null?

Reply:
I don't think a user should be allowed to pass in the null at all. If they don't have a config, they should use 
```
getCloudConfig(String zkAddress, String clusterName)
```
which would create a default for them. 

If they have a config, they should then call the new function, and they are responsible for having a correct config.

Reply:
From our offline discussion, we only need to create a default RealmAwareZkConnectionConfig when it is in multi ZK mode. Do it this way will initiate a  RealmAwareZkConnectionConfig object regardless we are in multiZk or not.
Let's keep the current approach.

Reply:
From our offline discussion, we only need to create a default RealmAwareZkConnectionConfig when it is in multi ZK mode. Do it this way will initiate a  RealmAwareZkConnectionConfig object regardless we are in multiZk or not.
Let's keep the current approach.